File: test_projects/calculus/src/parser.ts
import {
  ExpressionNode, // Import base class ExpressionNode
  NumberNode,
  VariableNode,
  BinaryOperationNode,
  UnaryOperationNode,
  FunctionCallNode,
  ConstantNode, // ADDED: ConstantNode
} from './expression'; // Import concrete classes
import {
  Expression, // Import interface
  BinaryOperatorType,
  UnaryOperatorType,
  FunctionNameType,
  ConstantType // ADDED: ConstantType
} from './expressionInterface'; // Import types and enums

// Simple token stream manager
class TokenStream {
  private tokens: string[];
  private currentIndex: number = 0;

  constructor(input: string) {
    this.tokens = this.tokenize(input);
  }

  // Basic tokenizer:
  // - Recognizes numbers (integers and floats)
  // - Recognizes variables/function names (alphabetic sequences)
  // - Recognizes operators and parentheses
  private tokenize(expression: string): string[] {
    // Regex for tokens:
    // 1. Numbers: \d+(\.\d*)?|\.\d+ (handles numbers like 123, 3.14, or .5)
    // 2. Variable/Function names: [a-zA-Z_]+
    // 3. Operators: \+|\-|\*|\/|\^|\(|\)|, (single character operators)
    const regex = /(\d+(\.\d*)?|\.\d+|[a-zA-Z_]+|\+|\-|\*|\/|\^|\(|\)|,)/g;
    let match;
    const result: string[] = [];
    while ((match = regex.exec(expression)) !== null) {
      if (match[0].trim() !== '') { // Ensure no empty matches from weird regex behavior
        result.push(match[0]);
      }
    }
    return result;
  }

  peek(): string | undefined {
    return this.tokens[this.currentIndex];
  }

  consume(): string {
    if (this.currentIndex >= this.tokens.length) {
      throw new Error('Unexpected end of input');
    }
    return this.tokens[this.currentIndex++];
  }

  // Consume a token only if it matches the expected type
  expect(expected: string): string {
    const token = this.consume();
    if (token !== expected) {
      throw new Error(`Expected '${expected}' but got '${token}'`);
    }
    return token;
  }

  // Check if current token matches and consume it
  match(expected: string): boolean {
    if (this.peek() === expected) {
      this.consume();
      return true;
    }
    return false;
  }

  isAtEnd(): boolean {
    return this.currentIndex >= this.tokens.length;
  }
}

export function parse(input: string): Expression {
  const stream = new TokenStream(input.trim());

  // parseFactor: Handles numbers, variables, function calls, and parenthesized expressions.
  function parseFactor(): ExpressionNode { // Changed return type to ExpressionNode
    const token = stream.peek();

    if (!token) {
      throw new Error('Unexpected end of input during factor parsing');
    }

    // Number (check for both integer and float formats, including leading dot for floats like .5)
    if (!isNaN(Number(token))) {
      // Ensure the token represents a valid number string (e.g., not just '.')
      if (token === '.' || (token.startsWith('.') && isNaN(Number('0' + token)))) {
        throw new Error(`Invalid number format: ${token}`);
      }
      return new NumberNode(Number(stream.consume()));
    }

    // Parenthesized expression
    if (stream.match('(')) {
      const expr = parseExpression();
      stream.expect(')');
      return expr;
    }

    // Variable or Function Call
    if (token.match(/^[a-zA-Z_]+$/)) {
      const name = stream.consume();

      // Check for constants like 'e' or 'pi' BEFORE checking for general variables
      const constants: ConstantType[] = ['e', 'pi'];
      if (constants.includes(name as ConstantType)) {
        return new ConstantNode(name as ConstantType);
      }

      if (stream.match('(')) { // It's a function call
        const args: ExpressionNode[] = []; // Changed type here to ExpressionNode[]
        if (stream.peek() === ')') { // If ')' immediately follows '(', it's an empty argument list
          throw new Error(`Missing argument`);
        }
        
        // Parse the first argument
        args.push(parseExpression());
        
        // As per current spec and tests, only single argument functions are expected.
        // If multiple arguments are encountered, it's an error for now.
        if (stream.peek() === ',') {
          throw new Error('Functions with multiple arguments are not supported by this parser yet.');
        }
        
        stream.expect(')'); // Expect closing parenthesis
        
        // Validate function name against allowed types
        const functionNames: FunctionNameType[] = [
          'log', 'ln', 'exp',
          'sin', 'cos', 'tan',
          'sec', 'csc', 'cot',
          'asin', 'acos', 'atan',
          'asec', 'acsc', 'acot',
          'sinh', 'cosh', 'tanh',
          'asinh', 'acosh', 'atanh'
        ];
        if (!functionNames.includes(name as FunctionNameType)) {
          throw new Error(`Unknown function: ${name}`);
        }
        return new FunctionCallNode(name as FunctionNameType, args as ExpressionNode[]); // Explicit cast for args
      } else { // It's a variable
        return new VariableNode(name);
      }
    }

    throw new Error(`Unexpected token at factor: ${token}`);
  }

  // parseUnary: Handles unary negation.
  function parseUnary(): ExpressionNode { // Changed return type to ExpressionNode
    if (stream.match('-')) {
      const operand = parseUnary(); // Allow nested unary operations (--x)
      return new UnaryOperationNode('negate', operand as ExpressionNode); // Explicit cast for operand
    }
    return parseFactor();
  }

  // parsePower: Handles exponentiation (right-associative).
  function parsePower(): ExpressionNode { // Changed return type to ExpressionNode
    let left = parseUnary();
    while (stream.match('^')) {
      const right = parsePower(); // Right-associative
      left = new BinaryOperationNode('power', left as ExpressionNode, right as ExpressionNode); // Explicit casts for left and right
    }
    return left;
  }

  // parseTerm: Handles multiplication and division (left-associative).
  function parseTerm(): ExpressionNode { // Changed return type to ExpressionNode
    let left = parsePower();
    while (stream.peek() === '*' || stream.peek() === '/') {
      const operatorToken = stream.consume();
      const right = parsePower();
      left = new BinaryOperationNode(
        operatorToken === '*' ? 'multiply' : 'divide',
        left as ExpressionNode, // Explicit cast for left
        right as ExpressionNode, // Explicit cast for right
      );
    }
    return left;
  }

  // parseExpression: Handles addition and subtraction (left-associative).
  function parseExpression(): ExpressionNode { // Changed return type to ExpressionNode
    let left = parseTerm();
    while (stream.peek() === '+' || stream.peek() === '-') {
      const operatorToken = stream.consume();
      const right = parseTerm();
      left = new BinaryOperationNode(
        operatorToken === '+' ? 'add' : 'subtract',
        left as ExpressionNode, // Explicit cast for left
        right as ExpressionNode, // Explicit cast for right
      );
    }
    return left;
  }

  const result = parseExpression();
  if (!stream.isAtEnd()) {
    throw new Error(`Unexpected tokens at end of input: ${stream.peek()}`);
  }
  return result;
}

File: test_project_compare/opus_4_calculus/parser.py
"""
Parser for converting string expressions into Expression AST.
"""
import re
from typing import List, Optional, Tuple
from expression import (
    Expression, Number, Variable, Add, Subtract, Multiply, Divide, 
    Power, Sin, Cos, Ln, Exp
)


class Token:
    """Represents a token in the expression."""
    
    def __init__(self, type: str, value: str):
        self.type = type
        self.value = value
    
    def __repr__(self):
        return f"Token({self.type}, {self.value})"


class Lexer:
    """Tokenizes mathematical expressions."""
    
    TOKEN_PATTERNS = [
        ('NUMBER', r'\d+\.?\d*'),
        ('VARIABLE', r'[a-zA-Z_]\w*'),
        ('PLUS', r'\+'),
        ('MINUS', r'-'),
        ('MULTIPLY', r'\*'),
        ('DIVIDE', r'/'),
        ('POWER', r'\^'),
        ('LPAREN', r'\('),
        ('RPAREN', r'\)'),
        ('COMMA', r','),
        ('WHITESPACE', r'\s+'),
    ]
    
    FUNCTIONS = {'sin', 'cos', 'ln', 'exp', 'log'}
    
    def __init__(self, expression: str):
        self.expression = expression
        self.position = 0
        self.tokens = []
        self._tokenize()
    
    def _tokenize(self):
        """Convert expression string into tokens."""
        while self.position < len(self.expression):
            match_found = False
            
            for token_type, pattern in self.TOKEN_PATTERNS:
                regex = re.compile(pattern)
                match = regex.match(self.expression, self.position)
                
                if match:
                    value = match.group(0)
                    if token_type != 'WHITESPACE':
                        # Check if it's a function
                        if token_type == 'VARIABLE' and value in self.FUNCTIONS:
                            self.tokens.append(Token('FUNCTION', value))
                        else:
                            self.tokens.append(Token(token_type, value))
                    
                    self.position = match.end()
                    match_found = True
                    break
            
            if not match_found:
                raise ValueError(f"Invalid character at position {self.position}: '{self.expression[self.position]}'")


class Parser:
    """Recursive descent parser for mathematical expressions."""
    
    def __init__(self, tokens: List[Token]):
        self.tokens = tokens
        self.position = 0
    
    def parse(self) -> Expression:
        """Parse tokens into an Expression AST."""
        if not self.tokens:
            raise ValueError("Empty expression")
        
        expr = self._parse_expression()
        
        if self.position < len(self.tokens):
            raise ValueError(f"Unexpected token: {self.tokens[self.position]}")
        
        return expr
    
    def _current_token(self) -> Optional[Token]:
        """Get current token without consuming it."""
        if self.position < len(self.tokens):
            return self.tokens[self.position]
        return None
    
    def _consume_token(self) -> Optional[Token]:
        """Consume and return current token."""
        if self.position < len(self.tokens):
            token = self.tokens[self.position]
            self.position += 1
            return token
        return None
    
    def _parse_expression(self) -> Expression:
        """Parse addition and subtraction (lowest precedence)."""
        left = self._parse_term()
        
        while self._current_token() and self._current_token().type in ('PLUS', 'MINUS'):
            op = self._consume_token()
            right = self._parse_term()
            
            if op.type == 'PLUS':
                left = Add(left, right)
            else:
                left = Subtract(left, right)
        
        return left
    
    def _parse_term(self) -> Expression:
        """Parse multiplication and division."""
        left = self._parse_factor()
        
        while self._current_token() and self._current_token().type in ('MULTIPLY', 'DIVIDE'):
            op = self._consume_token()
            right = self._parse_factor()
            
            if op.type == 'MULTIPLY':
                left = Multiply(left, right)
            else:
                left = Divide(left, right)
        
        return left
    
    def _parse_factor(self) -> Expression:
        """Parse power operations (right associative)."""
        base = self._parse_unary()
        
        if self._current_token() and self._current_token().type == 'POWER':
            self._consume_token()
            exponent = self._parse_factor()  # Right associative
            return Power(base, exponent)
        
        return base
    
    def _parse_unary(self) -> Expression:
        """Parse unary operations (negation)."""
        if self._current_token() and self._current_token().type == 'MINUS':
            self._consume_token()
            expr = self._parse_unary()
            return Multiply(Number(-1), expr)
        
        return self._parse_atom()
    
    def _parse_atom(self) -> Expression:
        """Parse atomic expressions (numbers, variables, functions, parentheses)."""
        token = self._current_token()
        
        if not token:
            raise ValueError("Unexpected end of expression")
        
        # Number
        if token.type == 'NUMBER':
            self._consume_token()
            return Number(float(token.value))
        
        # Variable
        elif token.type == 'VARIABLE':
            self._consume_token()
            return Variable(token.value)
        
        # Function
        elif token.type == 'FUNCTION':
            return self._parse_function()
        
        # Parentheses
        elif token.type == 'LPAREN':
            self._consume_token()
            expr = self._parse_expression()
            
            if not self._current_token() or self._current_token().type != 'RPAREN':
                raise ValueError("Missing closing parenthesis")
            
            self._consume_token()
            return expr
        
        else:
            raise ValueError(f"Unexpected token: {token}")
    
    def _parse_function(self) -> Expression:
        """Parse function calls."""
        func_token = self._consume_token()
        func_name = func_token.value
        
        if not self._current_token() or self._current_token().type != 'LPAREN':
            raise ValueError(f"Expected '(' after function {func_name}")
        
        self._consume_token()  # Consume '('
        
        # Parse argument
        arg = self._parse_expression()
        
        if not self._current_token() or self._current_token().type != 'RPAREN':
            raise ValueError(f"Expected ')' after function argument")
        
        self._consume_token()  # Consume ')'
        
        # Create appropriate function expression
        if func_name == 'sin':
            return Sin(arg)
        elif func_name == 'cos':
            return Cos(arg)
        elif func_name == 'ln' or func_name == 'log':
            return Ln(arg)
        elif func_name == 'exp':
            return Exp(arg)
        else:
            raise ValueError(f"Unknown function: {func_name}")


def parse_expression(expression_str: str) -> Expression:
    """Parse a string expression into an Expression AST."""
    # Handle special constants
    expression_str = expression_str.replace('e', str(math.e))
    expression_str = expression_str.replace('pi', str(math.pi))
    expression_str = expression_str.replace('π', str(math.pi))
    
    lexer = Lexer(expression_str)
    parser = Parser(lexer.tokens)
    return parser.parse()


# Import math for constants
import math